qualification/Jenkinsfile

```diff
@@ -0,0 +1,147 @@
+       * the Docker server available inside the container (for building a Docker
+       * image).
+       *
+       * --network=host: The docker container needs access to the internet to
+       * install packages and an interface to connect to the correlator network.
+       * This command passes all the host interfaces to the container to be
+       * used as required.
+       *
+       * --device=/dev/infiniband/rdma_cm and --device=/dev/infiniband/uverbs0:
+       * These flags pass the drivers required for ibverbs to the container.
+       * NOTE: The driver is not always ubverbs0, sometimes it is ubverbs1 or
+       * 2 etc. The correct once needs to be specified.
+       */
+      args '--gpus=all -v /var/run/docker.sock:/var/run/docker.sock --network=host -u=root --ulimit=memlock=-1 --device=/dev/infiniband/rdma_cm  --device=/dev/infiniband/uverbs0'
+    }
+  }
+
+  options {
+    timeout(time: 1, unit: 'HOURS')
+  }
+
+  stages {
+    stage('Install Python packages') {
+      steps {
+        sh 'pip install -r requirements.txt -r requirements-dev.txt'
+      }
+    }
+
+    stage('Install katgpucbf package') {
+      steps {
+        sh 'pip install --no-deps ".[test]" && pip check'
+      }
+    }
+
+    stage('Parallel stage') {
+      parallel {
+        stage('Compile C++ tools') {
+          steps {
+            // Make and compile tools.
+            dir('src/tools') {
+              sh 'make clean'
+              sh 'make -j'
+            }
+          }
+        }
+
+        /* This stage ensures that all the python style guidelines checks pass.
+         * This will catch if someone has committed to the repo without
+         * installing the required pre-commit hooks, or has bypassed them.
+         */
+        stage('Run pre-commit checks') {
+          steps {
+            sh 'pre-commit install'
+            // no-commit-to-branch complains if we are on the main branch
+            sh 'SKIP=no-commit-to-branch pre-commit run --all-files'
+          }
+        }
+      }
+    }
+
+    /* This stage executes the qualification tests on the latest docker image
+     * of katgpucbf that has been previously unit-tested, and generates a report of the
+     * qualification test results.
+     */
+    stage('Run qualification tests') {
+       steps {
+         script {
+           sh 'pip install -r qualification/requirements.txt'
+           sh "spead2_net_raw pytest -v qualification --image-override katgpucbf:harbor.sdp.kat.ac.za/cbf/katgpucbf:latest"
+           sh "qualification/report/generate_pdf.py report.json report.pdf"
+         }
+       }
+    }
+  }
+
+  /* This post stage is configured to always run at the end of this pipeline,
+   * regardless of the completion status. In this stage an email is sent to
+   * the specified address with details of the Jenkins job and a pdf report
+   * containing the qualification test results. The final step removes the workspace when
+   * the build is complete.
+   */
```

Comment:
Apart from emailing it, I'd suggest using the `publishHTML` step to store the report in Jenkins (take a look at what's done in the regular Jenkinsfile for the docs PDF). If I recall, `publishHTML` has an option to make it retain the report for all builds rather than just the latest one. That could be useful in future to trace regressions.

Reply:
This works great!